packages/react/src/components/Form/PinField.tsx

```diff
@@ -0,0 +1,157 @@
+
+import { FieldSize, FieldState } from './Helpers';
+
+type Values = string[];
+
+interface PinProps {
+  pattern?: RegExp,
+  autoFocus?: boolean;
+  secret?: boolean;
+  state?: FieldState;
+  onComplete?: (value) => void;
+}
+
+interface PinHookProps extends PinProps {
+  values?: Values;
+  onChange?: (values: Values) => void;
+}
+
+export interface PinFieldProps extends PinProps {
+  label?: string;
+  value?: string | number;
+  length?: number;
+  size?: FieldSize;
+  className?: string;
+}
+
+const PinField = ({
+  label,
+  value,
+  length = 4,
+  state,
+  size,
+  secret = false,
+  autoFocus = false,
+  pattern = DEFAULT_PATTERN,
+  className,
+  onComplete,
+}: PinFieldProps) => {
+  const containerClassNames = ['pin-field', size, className].join(' ');
+  const fieldClassNames = ['form-field', state].join(' ');
+  const labelClassNames = ['form-label', state].join(' ');
+
+  const defaultValues = Array.from({ length }).map((_, i) => value?.toString()[i] || '');
+  const [values, setValues] = useState(defaultValues);
+  const fields = usePinHook({
+    values,
+    autoFocus,
+    secret,
+    pattern,
+    state,
+    onChange: v => setValues(v),
+    onComplete,
+  });
+
+  return (
+    <div className="form-field-row">
+      {label && (
+        <label className={labelClassNames}>
+          {label}
+        </label>
+      )}
+      <div className={containerClassNames}>
+        {fields.map((fieldProps, index) => (
+          <div key={index} className={fieldClassNames}>
+            <input className="form-input" {...fieldProps} />
+          </div>
+        ))}
+      </div>
+    </div>
+  );
+};
+
+PinField.displayName = 'PinField';
+
+const usePinHook = ({
+  values,
+  autoFocus,
+  secret,
+  state,
+  pattern,
+  onChange: onChangeProp,
+  onComplete,
+}: PinHookProps = {}) => {
+  const disabled = state === 'disabled';
+  const error = state === 'error';
+
+  const fieldRefs = useRef<HTMLInputElement[]>(Array(values.length).fill(null));
+
+  const setFocus = useCallback((index = 0) => {
+    fieldRefs.current[index]?.focus();
+  }, []);
+
+  useEffect(() => {
+    if (autoFocus) setFocus();
+  }, [autoFocus, setFocus]);
+
+  const setFieldRef = useCallback((index: number) => (ref: HTMLInputElement) => {
+    fieldRefs.current[index] = ref;
+  }, []);
+
+  const onChange = useCallback((index: number) => (event: ChangeEvent<HTMLInputElement>) => {
+    let { value } = event.target;
+
+    value = value.trim().toUpperCase();
+
+    if (!pattern.test(value)) return;
+
+    const nextValues = values.slice();
+    nextValues[index] = value;
+    onChangeProp(nextValues);
+
+    if (value) {
+      if (!nextValues.includes('')) {
+        onComplete?.(nextValues.join(''));
+      }
+
+      if (index !== values.length - 1) {
+        if (error) {
+          const emptyFieldIndex = nextValues.findIndex((v) => !v);
+
+          if (emptyFieldIndex !== -1) {
+            setFocus(emptyFieldIndex);
+          }
```

Comment:
when there's an error focus would jump to the first empty field
but if we're on the last square, focus would not change to the first empty spot 

we don't need to add such logic right now, but can add it as an improvement later if there's a need for it

Reply:
So you're suggesting removing the whole "error focus" logic?

Reply:
Yes, let's leave just the minimal needed logic
- no special error handling
- no capitalization

UX features like these should come from requirements

Reply:
We should just be able to move to the next field when we add a character, move to the previous field when we delete a character, use Tab to switch between fields